File: gcs/src/test/scala/com/google/cloud/storage/contrib/nio/testing/FakeRpcBatch.scala
package com.google.cloud.storage.contrib.nio.testing

import java.util
import com.google.api.services.storage.model.StorageObject
import com.google.cloud.storage.spi.v1.{RpcBatch, StorageRpc}

class FakeRpcBatch(storageRpc: StorageRpc) extends RpcBatch {
  val deletions =
    collection.mutable.ListBuffer[(StorageObject, RpcBatch.Callback[Void], util.Map[StorageRpc.Option, _])]()

  def addDelete(
    storageObject: StorageObject,
    callback: RpcBatch.Callback[Void],
    options: util.Map[StorageRpc.Option, _]
  ): Unit = deletions.append((storageObject, callback, options))

  def addPatch(
    storageObject: StorageObject,
    callback: RpcBatch.Callback[StorageObject],
    options: util.Map[StorageRpc.Option, _]
  ): Unit = ???

  def addGet(
    storageObject: StorageObject,
    callback: RpcBatch.Callback[StorageObject],
    options: util.Map[StorageRpc.Option, _]
  ): Unit = ???

  def submit(): Unit = deletions.foreach { case (o, callback, options) =>
    storageRpc.delete(o, options)
    callback.onSuccess(null) // scalafix:ok
  }
}

File: gcs/src/test/java/com/google/cloud/storage/contrib/nio/testing/FakeRpcBatch.java
package com.google.cloud.storage.contrib.nio.testing;

import com.google.api.services.storage.model.StorageObject;
import com.google.cloud.storage.spi.v1.RpcBatch;
import com.google.cloud.storage.spi.v1.StorageRpc;
import scala.Tuple3;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;

public class FakeRpcBatch implements RpcBatch {

    protected StorageRpc storageRpc;
    protected Collection<scala.Tuple3<StorageObject, RpcBatch.Callback<Void>, Map<StorageRpc.Option, ?>>> deletions;

    public FakeRpcBatch(StorageRpc rpc) {
        storageRpc = rpc;
        deletions = new ArrayList<>();
    }

    @Override
    public void addDelete(StorageObject storageObject, RpcBatch.Callback<Void> callback, Map<StorageRpc.Option, ?> options) {
        deletions.add(new Tuple3<>(storageObject, callback, options));
    }

    @Override
    public void addPatch(StorageObject storageObject, RpcBatch.Callback<StorageObject> callback, Map<StorageRpc.Option, ?> options) {
        throw new UnsupportedOperationException("Not implemented");
    }

    @Override
    public void addGet(StorageObject storageObject, RpcBatch.Callback<StorageObject> callback, Map<StorageRpc.Option, ?> options) {
        throw new UnsupportedOperationException("Not implemented");
    }

    @Override
    public void submit() {
        deletions.forEach(tuple -> {
            storageRpc.delete(tuple._1(), tuple._3());
            tuple._2().onSuccess(null);
        });
    }
}